src/GeneratedHydrator/ClassGenerator/AllowedPropertiesOption.php

```diff
@@ -0,0 +1,155 @@
+     */
+    private $allowedProperties;
+
+    public function __construct(ReflectionClass $reflectedClass, array $options) {
+        $this->propertyNames = array_map(function($prop) {
+            return $prop->name;
+        }, $reflectedClass->getProperties());
+
+        $this->allowedProperties = $this->expandAllowedProperties($options);
+    }
+
+    /**
+     * Returns an array with properties as keys and hydrate/extract information
+     * as values.
+     *
+     * @param array $options
+     */
+    private function expandAllowedProperties($options)
+    {
+        $allowedProperties = [];
+
+        // Option was not given
+        if (! isset($options[static::OPTION_ALLOWED_PROPERTIES])) {
+            foreach ($this->propertyNames as $propertyName) {
+                $allowedProperties[$propertyName] = [
+                    'extract' => true,
+                    'hydrate' => true
+                ];
+            }
+
+            return $allowedProperties;
+        }
+
+        if (! is_array($options[static::OPTION_ALLOWED_PROPERTIES])) {
+            throw InvalidOptionException::valueNotArray(gettype($options[static::OPTION_ALLOWED_PROPERTIES]));
+        }
+
+        // Option was given
+        foreach ($options[static::OPTION_ALLOWED_PROPERTIES] as $k => $v) {
+            // simple format
+            if (is_int($k)) {
+                $this->makeSimpleFormat($k, $v, $allowedProperties);
+
+                continue;
+            }
+
+            // advanced format
+            if (is_string($k)) {
+                $this->makeAdvancedFormat($k, $v, $allowedProperties);
+            }
+        }
+
+        // Disable all properties which are not specified in the allowedProperties
+        foreach ($this->propertyNames as $propertyName) {
+            if (! in_array($propertyName, array_keys($allowedProperties))) {
+                $allowedProperties[$propertyName] = [
+                    'extract' => false,
+                    'hydrate' => false
+                ];
+            }
+        }
+
+        return $allowedProperties;
+    }
+
+    private function makeSimpleFormat($key, $value, &$allowedProperties) {
+        if (! is_string($value)) {
+            throw InvalidOptionException::invalidValueExpectedString(gettype($value), $key);
+        }
+
+        if (in_array($value, array_keys($allowedProperties))) {
+            throw InvalidOptionException::doubleProperty($value);
+        }
+
+        $allowedProperties[$value] = [
+            'extract' => true,
+            'hydrate' => true
+        ];
+    }
+
+    private function makeAdvancedFormat($key, $value, &$allowedProperties) {
+        if (! is_array($value)) {
+            throw InvalidOptionException::arrayExpected($k, gettype($value));
+        }
+
+        if (in_array($key, $allowedProperties)) {
+            throw InvalidOptionException::doubleProperty($value);
+        }
+
+        $validateOptionConfigurationKey = function($property, $array, $key) {
```

Comment:
This is because it's a multi dimensional array ... so the key/value of one level have a different meaning on the next level. Don't think we can do much about this